spot_driver_cpp/src/interfaces/rclcpp_tf_interface.cpp

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) 2023 Boston Dynamics AI Institute LLC. All rights reserved.
+
+#include <spot_driver_cpp/interfaces/rclcpp_tf_interface.hpp>
+
+namespace spot_ros2
+{
+  RclcppTfInterface::RclcppTfInterface(const std::shared_ptr<rclcpp::Node>& node)
+  : static_tf_broadcaster_{ node }
+  {
+  }
+
+  tl::expected<void, std::string> RclcppTfInterface::publishStaticTransforms(const std::vector<geometry_msgs::msg::TransformStamped>& transforms)
+  {
+    // Keep track of which transforms the static transform publisher is already publishing.
+    std::vector<geometry_msgs::msg::TransformStamped> new_transforms;
+    for (const auto& transform : transforms)
+    {
+      const auto frame_id_pair = std::make_pair(transform.header.frame_id, transform.child_frame_id);
+      // Do not publish transforms which have already previously been published.
```

Comment:
I don't think that's right.  I think the static transform broadcaster only remembers the last message.  So you have to publish all the transforms you want broadcast even if they've been previously broadcast.  Static transform broadcasts are costly though so you only want to publish a new one if it's a new transform is my understanding.

Reply:
After digging into this a bit it looks like [the rclcpp version](https://github.com/ros2/geometry2/blob/f2e7e1417ea599e719adbf08cd386fdca78e82ec/tf2_ros/src/static_transform_broadcaster.cpp) of the StaticTransformBroadcaster functions differently than [the rclpy version](https://github.com/ros2/geometry2/blob/rolling/tf2_ros_py/tf2_ros/static_transform_broadcaster.py).

In [the rclcpp version's sendTransform function](https://github.com/ros2/geometry2/blob/f2e7e1417ea599e719adbf08cd386fdca78e82ec/tf2_ros/src/static_transform_broadcaster.cpp#L52-L72), it has some logic to compare the new transform against previously-published transforms to determine if the stored list of previously-published transforms needs to be updated, and it always publishes all previously-published transforms when sendTransform is called.

I added a doc comment explaining what's going on behind the scenes here, and I updated the logic in the interface class to only call sendTransform if there's an entirely new transform to avoid unnecessarily broadcasting redundant static transforms.

Reply:
Where is the exploding head emojii in github reviews